File: shared/agoapp.cpp
#include <signal.h>
#include <stdexcept>
#include <boost/algorithm/string/case_conv.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/algorithm/string/join.hpp>



#include "agoapp.h"

#include "agoclient.h"
#include "options_validator.hpp"

using namespace qpid::types;
using namespace qpid::messaging;

using namespace agocontrol::log;
using namespace boost_options_validator;
namespace po = boost::program_options;
namespace fs = boost::filesystem;

namespace agocontrol {

/* Single instance of AgoApp which gets info on signals */
static AgoApp *app_instance;
static void static_sighandler(int signo) {
    if(app_instance != NULL)
        app_instance->sighandler(signo);
}


AgoApp::AgoApp(const char *appName_)
    : appName(appName_)
      , exit_signaled(false)
      , agoConnection(NULL)
{
    // Keep a "short" name too, trim leading 'Ago' from app name
    // This will be used for getConfigOption section names
    if(appName.find("Ago") == 0)
        appShortName = appName.substr(3);
    else
        appShortName = appName;

    boost::to_lower(appShortName);

    // Default to same config section as shortname, but this may be custom
    // if the original app did not use something "matching"
    appConfigSection = appShortName;
}

int AgoApp::parseCommandLine(int argc, const char **argv) {
    po::options_description options;
    options.add_options()
        ("help,h", "produce this help message")
        ("log-level", po::value<std::string>(),
         (std::string("Log level. Valid values are one of\n ") +
          boost::algorithm::join(log_container::getLevels(), ", ")).c_str())
        ("log-method", po::value<std::string>(),
         "Where to log. Valid values are one of\n console, syslog")
        ("log-syslog-facility", po::value<std::string>(),
         (std::string("Which syslog facility to log to. Valid values are on of\n ") +
          boost::algorithm::join(log_container::getSyslogFacilities(), ", ")).c_str())
        ("debug,d", po::bool_switch(),
         "Shortcut to set console logging with level DEBUG")
        ("trace,t", po::bool_switch(),
         "Shortcut to set console logging with level TRACE")
        ("config-dir", po::value<std::string>(),
         "Directory with configuration files")
        ("state-dir", po::value<std::string>(),
         "Directory with local state files")
        ;

    appCmdLineOptions(options);
    try {
        po::variables_map &vm (cli_vars);
        po::store(po::parse_command_line(argc, argv, options), vm);
        po::notify(vm);

        if(vm.count("config-dir")) {
            std::string dir = vm["config-dir"].as<std::string>();
            try {
                AgoClientInternal::setConfigDir(dir);
            } catch(const fs::filesystem_error& error) {
                std::cout << "Could not use " << dir << " as config-dir: "
                    << error.code().message()
                    << std::endl;
                return 1;
            }
        }

        if(vm.count("state-dir")) {
            std::string dir = vm["state-dir"].as<std::string>();
            try {
                AgoClientInternal::setLocalStateDir(dir);
            } catch(const fs::filesystem_error& error) {
                std::cout << "Could not use " << dir << " as state-dir: "
                    << error.code().message()
                    << std::endl;
                return 1;
            }
        }

        // Init dirs before anything else, so we at least show correct in help output
        if (vm.count("help")) {
            initDirectorys();
            std::cout << "usage: " << argv[0] << std::endl
                << options << std::endl
                << std::endl
                << "Paths:" << std::endl
                << "  Default config dir: "
                <<  BOOST_PP_STRINGIZE(DEFAULT_CONFDIR) << std::endl
                << "  Default state dir : "
                <<  BOOST_PP_STRINGIZE(DEFAULT_LOCALSTATEDIR) << std::endl
                << "  Active config dir : "
                << getConfigPath().string() << std::endl
                << "  Active state dir  : "
                << getLocalStatePath().string() << std::endl
                << std::endl
                << "System configuration file      : "
                << getConfigPath("conf.d/system.conf").string() << std::endl
                << "App-specific configuration file: "
                << getConfigPath("conf.d/" + appShortName + ".conf").string() << std::endl
                ;
            return 1;
        }

        options_validator<std::string>(vm, "log-method")
            ("console")("syslog")
            .validate();

        options_validator<std::string>(vm, "log-syslog-facility")
            (log_container::getSyslogFacilities())
            .validate();

        options_validator<std::string>(vm, "log-level")
            (log_container::getLevels())
            .validate();
    }
    catch(std::exception& e)
    {
        std::cout << "Failed to parse command line: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}


void AgoApp::setup() {
    setupLogging();
    setupAgoConnection();
    setupSignals();
    setupApp();
    setupIoThread();
}

void AgoApp::cleanup() {
    cleanupApp();
    cleanupAgoConnection();
    cleanupIoThread();

    // Global signal mapper
    app_instance = NULL;
}

void AgoApp::setupLogging() {
    log_container::initDefault();

    std::string level_str;
    if(cli_vars["debug"].as<bool>())
        level_str = "DEBUG";
    else if(cli_vars["trace"].as<bool>())
        level_str = "TRACE";
    else if(cli_vars.count("log-level"))
        level_str = cli_vars["log-level"].as<std::string>();
    else
        level_str = getConfigOption(appShortName.c_str(), "system", "log_level", "info");

    try {
        severity_level level = log_container::getLevel(level_str);
        log_container::setCurrentLevel(level);
    }catch(std::runtime_error &e) {
        throw ConfigurationError(e.what());
    }

    std::string method;
    if(cli_vars["debug"].as<bool>() || cli_vars["trace"].as<bool>())
        method = "console";
    else if(cli_vars.count("log-method"))
        method = cli_vars["log-method"].as<std::string>();
    else
        method = getConfigOption(appShortName.c_str(), "system", "log_method", "console");

    if(method == "syslog") {
        std::string facility_str;
        if(cli_vars.count("log-syslog-facility"))
            facility_str = cli_vars["log-syslog-facility"].as<std::string>();
        else
            facility_str = getConfigOption(appShortName.c_str(), "system", "syslog_facility", "local0");

        int facility = log_container::getFacility(facility_str);
        if(facility == -1) {
            throw ConfigurationError("Bad syslog facility '" + facility_str + "'");
        }

        log_container::setOutputSyslog(boost::to_lower_copy(appName), facility);
        AGO_DEBUG() << "Using syslog facility " << facility_str << ", level " << level_str;
    }
    else if(method == "console") {
        AGO_DEBUG() << "Using console log, level " << level_str;
    }
    else {
        throw ConfigurationError("Bad log_method '" + method + "'");
    }
}


void AgoApp::setupAgoConnection() {
    agoConnection = new AgoConnection(appShortName.c_str());
}

void AgoApp::cleanupAgoConnection() {
    if(agoConnection == NULL)
        return;

    delete agoConnection;
    agoConnection = NULL;
}

void AgoApp::addCommandHandler() {
    agoConnection->addHandler(boost::bind(&AgoApp::commandHandler, this, _1));
}

void AgoApp::addEventHandler() {
    agoConnection->addEventHandler(boost::bind(&AgoApp::eventHandler, this, _1, _2));
}




void AgoApp::setupSignals() {
    struct sigaction sa;
    sa.sa_handler = static_sighandler;
    sa.sa_flags = 0;
    //	sa.sa_flags = SA_RESTART;

    sigfillset(&sa.sa_mask);

    // static_sighandler calls global
    // static app_instance->sighandler
    app_instance = this;

    sigaction(SIGINT, &sa, NULL);
    sigaction(SIGQUIT, &sa, NULL);
}

void AgoApp::sighandler(int signo) {
    switch(signo) {
        case SIGINT:
        case SIGQUIT:
            AGO_DEBUG() << "Exit signal catched, shutting down";
            signalExit();
            break;

        default:
            AGO_WARNING() << "Unmapped signal " << signo << " received";
    }
}

void AgoApp::signalExit() {
    exit_signaled = true;

    // Dispatch signalExitThr call in separate thread, for application
    // to do more serious shutdown procedures
    boost::thread t(boost::bind(&AgoApp::doShutdown, this));
    t.detach();
}

void AgoApp::doShutdown() {
    try {
        if(agoConnection)
            agoConnection->shutdown();

    } catch(std::exception &e) {
    }
}



boost::asio::io_service& AgoApp::ioService() {
    if(!ioWork.get()) {
        // Enqueue work on the IO service to indicate we want it running
        ioWork = std::auto_ptr<boost::asio::io_service::work>(
                new boost::asio::io_service::work(ioService_)
                );
    }

    return ioService_;
}

void AgoApp::setupIoThread() {
    if(!ioWork.get()) {
        // No work job enqueued, skipping
        AGO_TRACE() << "Skipping IO thread, nothing scheduled";
        return;
    }

    // This thread will run until we're out of work.
    AGO_TRACE() << "Starting IO thread";
    ioThread = boost::thread(boost::bind(&boost::asio::io_service::run, &ioService_));
}

void AgoApp::cleanupIoThread(){
    if(!ioThread.joinable()) {
        AGO_TRACE() << "No IO thread alive";
        return;
    }

    AGO_TRACE() << "Resetting work & joining IO thread, waiting for it to exit";
    ioWork.reset();
    ioThread.join();
    AGO_TRACE() << "IO thread dead";
}




int AgoApp::appMain() {
    agoConnection->run();
    return 0;
}

int AgoApp::main(int argc, const char **argv) {
    try {
        if(parseCommandLine(argc, argv) != 0)
            return 1;
        setup();
    }catch(StartupError &e) {
        cleanup();
        return 1;
    }catch(ConfigurationError &e) {
        std::cerr << "Failed to start " << appName
            << " due to configuration error: "
            << e.what()
            << std::endl;

        cleanup();
        return 1;
    }

    AGO_INFO() << "Starting " << appName;

    int ret = this->appMain();

    AGO_DEBUG() << "Shutting down " << appName;

    cleanup();

    if(ret == 0)
        AGO_INFO() << "Exiting " << appName << "(code " << ret << ")";
    else
        AGO_WARNING() << "Exiting " << appName << "(code " << ret << ")";

    return ret;
}


}/*namespace agocontrol */

File: shared/agoclient/agoapp.py
import sys
import logging
import os.path
from logging.handlers import SysLogHandler
import argparse
import signal

import _logging
import _directories
import config
from agoconnection import AgoConnection

__all__ = ["AgoApp"]

class StartupError(Exception):
    pass
class ConfigurationError(Exception):
    pass

class AgoApp:
    """This is a base class for all Python AgoControl applications

    Each application needs to implement a class which extends the AgoApp class.
    Please see devices/example/ for basic usage examples.

    It must then provide the main entry point:

        if __name__ == "__main__":
            TheApp().main()

    """
    def __init__(self):
        self.app_name = self.__class__.__name__
        if self.app_name.find("Ago") == 0:
            self.app_short_name = self.app_name[3:].lower()
        else:
            self.app_short_name = self.app_name.lower()

        self.log = logging.getLogger(self.app_name)
        self.exit_signaled = False

    def parse_command_line(self, argv):
        """Parse the provided command line.

        This sets up the default options for all apps. If an application
        wants custom options, override the app_cmd_line_options method.

        Arguments are parsed by the argparse module, and the resulting Namespace
        object is available in self.args.

        Arguments:
            argv -- An array of parameters to parse, not including program name.
        """
        parser = argparse.ArgumentParser(add_help=False)

        LOG_LEVELS = ['TRACE', 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'FATAL']

        parser.add_argument('-h', '--help', action='store_true',
                help='show this help message and exit')

        parser.add_argument('--log-level', dest="log_level",
                help='Log level', choices = LOG_LEVELS)
        parser.add_argument('--log-method', dest="log_method",
                help='Where to log', choices = ['console', 'syslog'])

        facilities = SysLogHandler.facility_names.keys()
        facilities.sort()
        parser.add_argument('--log-syslog-facility', dest="syslog_facility",
                help='Which syslog facility to log to.',
                choices = facilities)

        parser.add_argument('-d', '--debug', action='store_true',
                help='Shortcut to set console logging with level DEBUG')
        parser.add_argument('-t', '--trace', action='store_true',
                help='Shortcut to set console logging with level TRACE')

        parser.add_argument('--config-dir', dest="config_dir",
                help='Directory with configuration files')
        parser.add_argument('--state-dir', dest="state_dir",
                help='Directory with local state files')

        self.app_cmd_line_options(parser)

        # If parsing fails, this will print error and exit
        args = parser.parse_args(argv[1:])

        if args.config_dir:
            config.set_config_dir(args.config_dir)

        if args.state_dir:
            config.set_localstate_dir(args.state_dir)

        if args.help:
            config.init_directories()

            parser.print_help()

            print
            print "Paths"
            print "  Default config dir: %s" % _directories.DEFAULT_CONFDIR
            print "  Default state dir : %s" % _directories.DEFAULT_LOCALSTATEDIR
            print "  Active config dir : %s" % config.CONFDIR
            print "  Active state dir  : %s" % config.LOCALSTATEDIR
            print
            print "System configuration file      : %s" % config.get_config_path('conf.d/system.conf')
            print "App-specific configuration file: %s" % config.get_config_path('conf.d/%s.conf'% self.app_short_name)
            print

            return False

        self.args = args
        return True

    def app_cmd_line_options(self, parser):
        """Override this to add your own command line options

        Arguments:
            parser -- An instance of argparse.ArgumentParser which the implementation should
                      add options to.
        """
        pass

    def setup(self):
        """Execute initial setup.

        This is done after command line arguments have been parsed.
        App specific setup should be done in setup_app method.
        """
        self.setup_logging()
        self.setup_connection()
        self.setup_signals()
        self.setup_app()

    def cleanup(self):
        """Cleanup after shutdown.

        Note that this may be called even if corresponding setup call
        have not been executed, implementations must ensure that they
        can handle this."""
        self.log.trace("Cleaning up")
        self.cleanup_app()
        self.cleanup_connection()

    def setup_logging(self):
        root = logging.getLogger()
        # Find log level..
        if self.args.trace:
            lvl_name = "TRACE"
        elif self.args.debug:
            lvl_name = "DEBUG"
        elif self.args.log_level:
            lvl_name = self.args.log_level
        else:
            lvl_name = self.get_config_option("log_level", "INFO",
                    section=[None, "system"])

            if lvl_name.upper() not in logging._levelNames:
                raise ConfigurationError("Invalid log_level %s" % lvl_name)

        # ..and set it
        lvl = logging._levelNames[lvl_name.upper()]
        root.setLevel(lvl)

        # Find log method..
        if self.args.trace or self.args.debug:
            log_method = 'console'
        elif self.args.log_method:
            log_method = self.args.log_method
        else:
            log_method = self.get_config_option("log_method", "console",
                    section=[None, "system"])

            if log_method not in ['console', 'syslog']:
                raise ConfigurationError("Invalid log_method %s" % log_method)

        # ..and set it
        if log_method == 'console':
            self.log_handler = logging.StreamHandler(None)
            self.log_formatter = logging.Formatter("%(asctime)-15s %(name)-10s %(levelname)-5s %(message)s")
        elif log_method == 'syslog':
            if self.args.syslog_facility:
                syslog_fac = self.args.syslog_facility
            else:
                syslog_fac = self.get_config_option("syslog_facility",
                        "local0", section = [None,"system"])

                if syslog_fac not in SysLogHandler.facility_names:
                    raise ConfigurationError("Invalid syslog_facility %s" % syslog_fac)

            # Try to autodetect OS syslog unix socket
            for f in ["/dev/log", "/var/run/log"]:
                if os.path.exists(f):
                    syslog_path = f
                    break

            self.log_handler = SysLogHandler(
                    address = syslog_path,
                    facility = SysLogHandler.facility_names[syslog_fac])

            self.log_formatter = logging.Formatter(self.app_name.lower() + " %(name)-10s %(levelname)-5s %(message)s")

        self.log_handler.setFormatter(self.log_formatter)

        # Forcibly limit QPID logging to INFO
        logging.getLogger('qpid').setLevel(max(root.level, logging.INFO))

        root.addHandler(self.log_handler)

    def setup_connection(self):
        """Create an AgoConnection instance, assigned to self.connection"""
        self.connection = AgoConnection(self.app_short_name)

    def cleanup_connection(self):
        """Shutdown and clean up our AgoConnection instance"""
        if self.connection:
            self.connection.shutdown()
            self.connection = None

    def setup_signals(self):
        """Setup signal handlers"""
        signal.signal(signal.SIGINT, self._sighandler)
        signal.signal(signal.SIGQUIT, self._sighandler)

    def _sighandler(self, signal, frame):
        """Internal method called when catched signals are received"""
        self.log.debug("Exit signal catched, shutting down")
        self.signal_exit()

    def signal_exit(self):
        """Call this to begin shutdown procedures.

        This can be called by the application if it wants to shut down the app.
        """
        self.exit_signaled = True
        self._do_shutdown()

    def is_exit_signaled(self):
        """Check if we have been asked to exit"""
        return self.exit_signaled

    def _do_shutdown(self):
        if self.connection:
            self.connection.shutdown()

    def setup_app(self):
        """This should be overriden by the application to setup app specifics"""
        pass

    def cleanup_app(self):
        """This should be overriden by the application to cleanup app specifics"""
        pass

    def main(self, argv=None):
        """Main entrypoint, called by the application

        Arguments:
            argv -- Command line arguments. Defaults to sys.argv if not set.
        """
        if not argv:
            argv = sys.argv

        ret = self._main(argv)
        sys.exit(ret)

    def _main(self, argv):
        """Internal main function.

        This is where we launch the app. It will:
            - parse command line
            - setup
            - run app main loop [forever]
            - cleanup
            - exit

        Arguments:
            argv -- Command line arguments, including script name in [0].

        Returns:
            OS Exit code

        """
        if not self.parse_command_line(argv):
            return 1

        try:
            self.setup()
        except StartupError:
            self.cleanup()
            return 1

        except ConfigurationError,e:
            self.log.error("Failed to start %s due to configuration error: %s",
                    self.app_name, e.message)
            self.cleanup()
            return 1

        try:
            self.log.info("Starting %s", self.app_name)
            ret = self.app_main()
            self.log.debug("Shutting down %s", self.app_name)

            self.cleanup()

            if ret == 0:
                self.log.info("Exiting %s", self.app_name)
            else:
                self.log.warn("Exiting %s (code %d)", self.app_name, ret)
        except:
            self.log.critical("Unhandled exception, crashing",
                    exc_info=True)

            return 1


    def app_main(self):
        """Main entry point for application.

        By default this calls the run method of the AgoConnection object, which will
        block until shut down.

        This CAN be overriden, but generally it is sufficient to do your setup in
        setup_app.
        """

        self.connection.run()
        return 0


    def get_config_option(self, option, default_value=None, section=None, app=None):
        """Read a config option from the configuration subsystem.

        The system is based on per-app configuration files, which has sections
        and options.

        Note that "option not set" means not set, or empty value!

        Arguments:
            option -- The name of the option to retreive

            default_value -- If the option can not be found in any of the specified
                sections, fall back to this value.

            section -- A string section to look for the option in, or an iterable
                with multiple sections to try in the defined order, in case the option
                was not set.

                If not set, it defaults to the applications short name,
                which is the class name in lower-case, with the leading "Ago" removed.
                (AgoExample becomes example).
                If you want to use the default value, but fall back on other, you
                can use None:

                    section = [None, 'system']

                This will look primarly in the default section, falling back to the
                'system' section.

            app -- A string identifying the configuration storage unit to look in.
                Can also be an iterable with multiple units to look at. If the option
                was not found in any of the sections specified, the next available
                'app' value will be tried.

                If not set, it defaults to the same value as section.
                If the list contains None, this is too replaced by the default section
                value (NOT the value of the passed section argument).


        Returns:
            A unicode object with the value found in the data store, if found.
            If not found, default_value is passed through unmodified.
        """
        if section is None:
            section = self.app_short_name

        if app is None:
            app = section

        config._iterable_replace_none(section, self.app_short_name)
        config._iterable_replace_none(app, self.app_short_name)

        return config.get_config_option(section, option, default_value, app)


    def set_config_option(self, option, value, section=None, app=None):
        """Write a config option to the configuration subsystem.

        The system is based on per-app configuration files, which has sections
        and options.

        Arguments:
            option -- The name of the option to set

            value -- The value of the option.

            section -- A string section in which to store the option in.
                If not set, it defaults to the applications short name,
                which is the class name in lower-case, with the leading "Ago" removed.
                (AgoExample becomes example).

            app -- A string identifying the configuration storage unit to store in.
                If omited, it defaults to the same as section.

        Returns:
            True if succesfully stored, False otherwise.
            Please refer to the error log for failure indication.

        """
        if section is None:
            section = self.app_short_name

        return config.set_config_option(section, option, value, app)

